include/sys_clock.h

```diff
@@ -1,248 +1,109 @@
-
-static inline int sys_clock_hw_cycles_per_sec(void)
-{
-#if defined(CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME)
-	return z_clock_hw_cycles_per_sec_runtime_get();
-#else
-	return CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC;
-#endif
-}
-
-/* Note that some systems with comparatively slow cycle counters
- * experience precision loss when doing math like this.  In the
- * general case it is not correct that "cycles" are much faster than
- * "ticks".
- */
-static inline int sys_clock_hw_cycles_per_tick(void)
-{
-#ifdef CONFIG_SYS_CLOCK_EXISTS
-	return sys_clock_hw_cycles_per_sec() / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
-#else
-	return 1; /* Just to avoid a division by zero */
-#endif
-}
-
-#if defined(CONFIG_SYS_CLOCK_EXISTS) && \
-	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC == 0)
-#error "SYS_CLOCK_HW_CYCLES_PER_SEC must be non-zero!"
-#endif
-
-/* number of nsec per usec */
-#define NSEC_PER_USEC 1000U
-
-/* number of microseconds per millisecond */
-#define USEC_PER_MSEC 1000U
-
-/* number of milliseconds per second */
-#define MSEC_PER_SEC 1000U
-
-/* number of microseconds per second */
-#define USEC_PER_SEC ((USEC_PER_MSEC) * (MSEC_PER_SEC))
-
-/* number of nanoseconds per second */
-#define NSEC_PER_SEC ((NSEC_PER_USEC) * (USEC_PER_MSEC) * (MSEC_PER_SEC))
-
-
-/* kernel clocks */
-
-/*
- * We default to using 64-bit intermediates in timescale conversions,
- * but if the HW timer cycles/sec, ticks/sec and ms/sec are all known
- * to be nicely related, then we can cheat with 32 bits instead.
- */
-
-#ifdef CONFIG_SYS_CLOCK_EXISTS
-
-#if defined(CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME) || \
-	(MSEC_PER_SEC % CONFIG_SYS_CLOCK_TICKS_PER_SEC) || \
-	(CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC % CONFIG_SYS_CLOCK_TICKS_PER_SEC)
-#define _NEED_PRECISE_TICK_MS_CONVERSION
-#endif
-
+#ifdef __cplusplus
+extern "C" {
 #endif
 
-static ALWAYS_INLINE s32_t z_ms_to_ticks(s32_t ms)
-{
-#ifdef CONFIG_SYS_CLOCK_EXISTS
-
-#ifdef _NEED_PRECISE_TICK_MS_CONVERSION
-	int cyc = sys_clock_hw_cycles_per_sec();
-
-	/* use 64-bit math to keep precision */
-	return (s32_t)ceiling_fraction((s64_t)ms * cyc,
-		((s64_t)MSEC_PER_SEC * cyc) / CONFIG_SYS_CLOCK_TICKS_PER_SEC);
+#ifdef CONFIG_SYS_TIMEOUT_64BIT
+typedef u64_t k_ticks_t;
 #else
-	/* simple division keeps precision */
-	s32_t ms_per_tick = MSEC_PER_SEC / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
-
-	return (s32_t)ceiling_fraction(ms, ms_per_tick);
+typedef u32_t k_ticks_t;
 #endif
 
+#ifdef CONFIG_SYS_TIMEOUT_LEGACY_API
+/* Fallback API where timeouts are still 32 bit millisecond counts */
+typedef k_ticks_t k_timeout_t;
+#define K_TIMEOUT_TICKS(t) (t)
+#define K_TIMEOUT_GET(t) (t)
+#define K_TIMEOUT_EQ(a, b) ((a) == (b))
 #else
-	__ASSERT(ms == 0, "ms not zero");
-	return 0;
+/* New API going forward: k_timeout_t is an opaque struct */
+typedef struct { k_ticks_t ticks; } k_timeout_t;
+#define K_TIMEOUT_TICKS(t) ((k_timeout_t){ (k_ticks_t)t })
+#define K_TIMEOUT_GET(t) ((t).ticks)
+#define K_TIMEOUT_EQ(a, b) ((a).ticks == (b).ticks)
```

Comment:
we should add macros for adding and substracting.
Consider following example: https://github.com/zephyrproject-rtos/zephyr/blob/477399627df082c758cfbdbdaec9c92a42a0af7b/subsys/net/buf.c#L319

diff is given in milliseconds so following operation would look like:
```
timeout = K_TIMEOUT_TICKS(TIMEOUT_GET(timeout) - MIN(K_TIMEOUT_GET(timeout), K_TIMEOUT_GET(K_MSEC(diff))))
```
which looks a bit complicated.

There could be macros like `K_TIMEOUT_SUB(timeout, raw_val)` and `K_TIMEOUT_ADD(timeout, raw_val)` which could be used in that case. Additionally, I'm missing macros for converting time to `k_ticks_t`, there is only conversion to `k_timeout_t`. In that particular case `K_TICKS_MSEC` would come handy for something like:
```
MIN(K_TIMEOUT_GET(timeout), K_TICKS_MSEC(diff))
```
After adding those macros operation would look:
```
K_TIMEOUT_SUB(timeout, MIN(K_TIMEOUT_GET(timeout), K_TICKS_MSEC(diff)));
```